client/controller.js

```diff
@@ -50,6 +55,8 @@ export function makeLoggedOutLayout( context, next ) {
  * divs.
  */
 export function clientRouter( route, ...middlewares ) {
+	routes.push( route );
+	routeMatcher = pathToRegexp( routes );
```

Comment:
Instead of tracking this stuff here ourselves, what about considering this a change in UI state, and hence unmounting as a side effect in `state/ui/actions` if the previous section was isomorphic, while the new one isn't? You could use the `isSectionIsomorphic()` selector found in one of the PRs we had to revert.